### PR TITLE
Enable workspace and organization Health Assessment (drift detection) setting management

### PR DESCRIPTION
**Pull request by rexredinger**
🚨 This has been updated to support and test against the settings changes described [here](https://github.com/hashicorp/atlas/pull/13300) 🚨

## Description

Allow go-tfe to access and manage drift detection setting for workspaces in TFC.

## External links

- [API documentation](https://www.terraform.io/cloud-docs/api-docs/workspaces#request-body)
-[ tfe provider PR ](https://github.com/hashicorp/terraform-provider-tfe/pull/550)

## Output from tests
Including output from tests may require access to a TFE instance. Ignore this section if you have no environment to test against.

```
$ go test -run TestWorkspacesUpdate -v ./... -tags=integration
=== RUN   TestWorkspacesUpdate
=== RUN   TestWorkspacesUpdate/when_updating_a_subset_of_values
=== RUN   TestWorkspacesUpdate/with_valid_options
=== RUN   TestWorkspacesUpdate/when_options_includes_both_an_operations_value_and_an_enforcement_mode_value
=== RUN   TestWorkspacesUpdate/when_'agent'_execution_mode_is_specified_without_an_an_agent_pool_ID
=== RUN   TestWorkspacesUpdate/when_an_error_is_returned_from_the_api
=== RUN   TestWorkspacesUpdate/when_options_has_an_invalid_name
=== RUN   TestWorkspacesUpdate/when_options_has_an_invalid_organization
=== RUN   TestWorkspacesUpdate/when_options_include_trigger-patterns_(behind_a_feature_flag)
=== RUN   TestWorkspacesUpdate/when_options_include_both_trigger-patterns_and_trigger-paths_error_is_returned
=== RUN   TestWorkspacesUpdate/when_options_include_trigger-patterns_populated_and_empty_trigger-paths_workspace_is_updated
=== RUN   TestWorkspacesUpdate/when_options_include_VCSRepo_tags-regex
    helper_test.go:633: Export a valid OAUTH_CLIENT_GITHUB_TOKEN before running this test!
=== RUN   TestWorkspacesUpdate/when_options_include_tags-regex_and_file-triggers-enabled_is_true_an_error_is_returned
=== RUN   TestWorkspacesUpdate/when_options_include_both_non-empty_tags-regex_and_file-triggers-enabled_an_error_is_returned
=== RUN   TestWorkspacesUpdate/when_options_include_both_tags-regex_and_trigger-prefixes_an_error_is_returned
=== RUN   TestWorkspacesUpdate/when_options_include_both_tags-regex_and_trigger-patterns_error_is_returned
--- PASS: TestWorkspacesUpdate (49.53s)
    --- PASS: TestWorkspacesUpdate/when_updating_a_subset_of_values (2.25s)
    --- PASS: TestWorkspacesUpdate/with_valid_options (4.49s)
    --- PASS: TestWorkspacesUpdate/when_options_includes_both_an_operations_value_and_an_enforcement_mode_value (0.00s)
    --- PASS: TestWorkspacesUpdate/when_'agent'_execution_mode_is_specified_without_an_an_agent_pool_ID (0.00s)
    --- PASS: TestWorkspacesUpdate/when_an_error_is_returned_from_the_api (2.27s)
    --- PASS: TestWorkspacesUpdate/when_options_has_an_invalid_name (0.00s)
    --- PASS: TestWorkspacesUpdate/when_options_has_an_invalid_organization (0.00s)
    --- PASS: TestWorkspacesUpdate/when_options_include_trigger-patterns_(behind_a_feature_flag) (13.19s)
    --- PASS: TestWorkspacesUpdate/when_options_include_both_trigger-patterns_and_trigger-paths_error_is_returned (0.00s)
    --- PASS: TestWorkspacesUpdate/when_options_include_trigger-patterns_populated_and_empty_trigger-paths_workspace_is_updated (12.40s)
    --- SKIP: TestWorkspacesUpdate/when_options_include_VCSRepo_tags-regex (5.24s)
    --- PASS: TestWorkspacesUpdate/when_options_include_tags-regex_and_file-triggers-enabled_is_true_an_error_is_returned (0.00s)
    --- PASS: TestWorkspacesUpdate/when_options_include_both_non-empty_tags-regex_and_file-triggers-enabled_an_error_is_returned (0.00s)
    --- PASS: TestWorkspacesUpdate/when_options_include_both_tags-regex_and_trigger-prefixes_an_error_is_returned (0.00s)
    --- PASS: TestWorkspacesUpdate/when_options_include_both_tags-regex_and_trigger-patterns_error_is_returned (0.00s)
=== RUN   TestWorkspacesUpdateByID
=== RUN   TestWorkspacesUpdateByID/when_updating_a_subset_of_values
=== RUN   TestWorkspacesUpdateByID/with_valid_options
=== RUN   TestWorkspacesUpdateByID/when_an_error_is_returned_from_the_api
=== RUN   TestWorkspacesUpdateByID/without_a_valid_workspace_ID
--- PASS: TestWorkspacesUpdateByID (19.45s)
    --- PASS: TestWorkspacesUpdateByID/when_updating_a_subset_of_values (2.46s)
    --- PASS: TestWorkspacesUpdateByID/with_valid_options (4.15s)
    --- PASS: TestWorkspacesUpdateByID/when_an_error_is_returned_from_the_api (2.22s)
    --- PASS: TestWorkspacesUpdateByID/without_a_valid_workspace_ID (0.00s)
PASS
ok  	github.com/hashicorp/go-tfe	69.204s
?   	github.com/hashicorp/go-tfe/examples/organizations	[no test files]
?   	github.com/hashicorp/go-tfe/examples/workspaces	[no test files]
?   	github.com/hashicorp/go-tfe/mocks	[no test files]
```

```
$ go test -run TestWorkspacesRead -v ./... -tags=integration
=== RUN   TestWorkspacesRead
=== RUN   TestWorkspacesRead/when_the_workspace_exists
=== RUN   TestWorkspacesRead/when_the_workspace_does_not_exist
=== RUN   TestWorkspacesRead/when_the_organization_does_not_exist
=== RUN   TestWorkspacesRead/without_a_valid_organization
=== RUN   TestWorkspacesRead/without_a_valid_workspace
--- PASS: TestWorkspacesRead (1.35s)
    --- PASS: TestWorkspacesRead/when_the_workspace_exists (0.13s)
    --- PASS: TestWorkspacesRead/when_the_workspace_does_not_exist (0.08s)
    --- PASS: TestWorkspacesRead/when_the_organization_does_not_exist (0.08s)
    --- PASS: TestWorkspacesRead/without_a_valid_organization (0.00s)
    --- PASS: TestWorkspacesRead/without_a_valid_workspace (0.00s)
=== RUN   TestWorkspacesReadWithOptions
=== RUN   TestWorkspacesReadWithOptions/when_options_to_include_resource
--- PASS: TestWorkspacesReadWithOptions (4.62s)
    --- PASS: TestWorkspacesReadWithOptions/when_options_to_include_resource (0.65s)
PASS
ok  	github.com/hashicorp/go-tfe	5.974s
?   	github.com/hashicorp/go-tfe/examples/organizations	[no test files]
?   	github.com/hashicorp/go-tfe/examples/workspaces	[no test files]
?   	github.com/hashicorp/go-tfe/mocks	[no test files]
```

```
$ go test -run TestWorkspacesCreate/with_valid_options -v ./... -tags=integration
=== RUN   TestWorkspacesCreate
=== RUN   TestWorkspacesCreate/with_valid_options
--- PASS: TestWorkspacesCreate (11.94s)
    --- PASS: TestWorkspacesCreate/with_valid_options (4.28s)
PASS
ok  	github.com/hashicorp/go-tfe	12.250s
?   	github.com/hashicorp/go-tfe/examples/organizations	[no test files]
?   	github.com/hashicorp/go-tfe/examples/workspaces	[no test files]
?   	github.com/hashicorp/go-tfe/mocks	[no test files]
```

```
$ go test -run TestOrganizationsCreate -v ./... -tags=integration
=== RUN   TestOrganizationsCreate
=== RUN   TestOrganizationsCreate/with_valid_options
=== RUN   TestOrganizationsCreate/when_no_email_is_provided
=== RUN   TestOrganizationsCreate/when_no_name_is_provided
=== RUN   TestOrganizationsCreate/with_invalid_name
--- PASS: TestOrganizationsCreate (6.67s)
    --- PASS: TestOrganizationsCreate/with_valid_options (4.48s)
    --- PASS: TestOrganizationsCreate/when_no_email_is_provided (0.00s)
    --- PASS: TestOrganizationsCreate/when_no_name_is_provided (0.00s)
    --- PASS: TestOrganizationsCreate/with_invalid_name (0.00s)
PASS
ok  	github.com/hashicorp/go-tfe	6.888s
?   	github.com/hashicorp/go-tfe/examples/organizations	[no test files]
?   	github.com/hashicorp/go-tfe/examples/workspaces	[no test files]
?   	github.com/hashicorp/go-tfe/mocks	[no test files]
```

```
go test -run TestOrganization_Unmarshall -v ./... -tags=integration
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/go-tfe	0.209s [no tests to run]
?   	github.com/hashicorp/go-tfe/examples/organizations	[no test files]
?   	github.com/hashicorp/go-tfe/examples/workspaces	[no test files]
?   	github.com/hashicorp/go-tfe/mocks	[no test files]
```

### CODE REVIEW

File: workspace.go [hunk 1]
```diff
@@ -349,6 +350,10 @@ type WorkspaceCreateOptions struct {
 	// A list of tags to attach to the workspace. If the tag does not already
 	// exist, it is created and added to the workspace.
 	Tags []*Tag `jsonapi:"relation,tags,omitempty"`
+
+	// Optional: Whether to enabled drift detection for the workspace.
```
Comment (annawinkler) [inline workspace.go:354]:
```suggestion
	// Optional: Whether to enable drift detection for the workspace.
```

File: workspace.go [hunk 2]
```diff
@@ -447,6 +452,10 @@ type WorkspaceUpdateOptions struct {
 	// the environment when multiple environments exist within the same
 	// repository.
 	WorkingDirectory *string `jsonapi:"attr,working-directory,omitempty"`
+
+	// Optional: Whether to enabled drift detection for the workspace.
```
Comment (annawinkler) [inline workspace.go:456]:
```suggestion
	// Optional: Whether to enable drift detection for the workspace.
```

File: workspace.go [hunk 3]
```diff
@@ -349,6 +350,10 @@ type WorkspaceCreateOptions struct {
 	// A list of tags to attach to the workspace. If the tag does not already
 	// exist, it is created and added to the workspace.
 	Tags []*Tag `jsonapi:"relation,tags,omitempty"`
+
+	// Optional: Whether to enabled drift detection for the workspace.
+	// Requires remote execution mode, TFCB entitlement, and a valid agent pool to work
+	DriftDetection *bool `jsonapi:"attr,drift-detection,omitempty"`
```
Comment (annawinkler) [inline workspace.go:356]: It looks like the attributes are listed in alphabetical order (after the type). Could you move this new attribute to maintain that ordering?

File: workspace.go [hunk 4]
```diff
@@ -447,6 +452,10 @@ type WorkspaceUpdateOptions struct {
 	// the environment when multiple environments exist within the same
 	// repository.
 	WorkingDirectory *string `jsonapi:"attr,working-directory,omitempty"`
+
+	// Optional: Whether to enabled drift detection for the workspace.
+	// Requires remote execution mode, TFCB entitlement, and a valid agent pool to work
+	DriftDetection *bool `jsonapi:"attr,drift-detection,omitempty"`
```
Comment (annawinkler) [inline workspace.go:458]: Same suggestion here regarding ordering of the attributes.

**Review by annawinkler — CHANGES_REQUESTED** [review]
Thanks for the PR! Could you also include a Changelog update?

**Comment by elijah-roberts** [comment]
Super interested in this feature. @annawinkler is there anything else that needs to be added to get this PR merged?

Are you waiting for the tests to run successfully? @rexredinger  any thoughts on these? Not sure if it is a flaky test, or an actual config error.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,7 @@
 # Unreleased
 
+* Add support for managing workspace drift detection setting [#462](https://github.com/hashicorp/go-tfe/pull/462)
```
Comment (annawinkler) [inline CHANGELOG.md:3]: It looks like this PR adds support for enforcing assessments in an organization and workspace. Could you update the changelog entry to reflect that?

File: organization.go [hunk 6]
```diff
@@ -158,6 +159,9 @@ type OrganizationCreateOptions struct {
 
 	// Optional: SendPassingStatusesForUntriggeredSpeculativePlans toggles behavior of untriggered speculative plans to send status updates to version control systems like GitHub.
 	SendPassingStatusesForUntriggeredSpeculativePlans *bool `jsonapi:"attr,send-passing-statuses-for-untriggered-speculative-plans,omitempty"`
+
+	// Optional: AssessmentsEnforced toggles whether health assessment enablement is enforced across all assessable workspaces or if the decission to enabled health assessments is delegated to the workspace setting AssessmentsEnabled.
```
Comment (annawinkler) [inline organization.go:163]:
```suggestion
	// Optional: AssessmentsEnforced toggles whether health assessment enablement is enforced across all assessable workspaces or if the decision to enabled health assessments is delegated to the workspace setting AssessmentsEnabled.
```

Could you clarify what an assessable workspace is?

File: organization.go [hunk 7]
```diff
@@ -191,6 +195,9 @@ type OrganizationUpdateOptions struct {
 
 	// SendPassingStatusesForUntriggeredSpeculativePlans toggles behavior of untriggered speculative plans to send status updates to version control systems like GitHub.
 	SendPassingStatusesForUntriggeredSpeculativePlans *bool `jsonapi:"attr,send-passing-statuses-for-untriggered-speculative-plans,omitempty"`
+
+	// Optional: AssessmentsEnforced toggles whether health assessment enablement is enforced across all assessable workspaces or if the decission to enabled health assessments is delegated to the workspace setting AssessmentsEnabled.
```
Comment (annawinkler) [inline organization.go:199]:
```suggestion
	// Optional: AssessmentsEnforced toggles whether health assessment enablement is enforced across all assessable workspaces or if the decision to enabled health assessments is delegated to the workspace setting AssessmentsEnabled.
```

**Review by annawinkler — CHANGES_REQUESTED** [review]
🥇 This is looking really good! I have a few small suggestions and then I think this PR will be ready to go after we've merged & built go-tfe with that PR.

File: workspace_integration_test.go [hunk 8]
```diff
@@ -671,12 +673,13 @@ func TestWorkspacesUpdate(t *testing.T) {
 
 	t.Run("when updating a subset of values", func(t *testing.T) {
```
Comment (annawinkler) [inline workspace_integration_test.go:674]: We'll need to upgrade the organization for this test to pass, so adding something like this:

```
upgradeOrganizationSubscription(t, client, orgTest)
```

On or around line 670 would work.

**Comment by rexredinger** [comment]
Updated workspace test with the `upgradeOrganizationSubscription`
and got this happy result in staging.
```
go test -run TestWorkspacesUpdate/when_updating_a_subset_of_values -v ./... -tags=integration
=== RUN   TestWorkspacesUpdate
=== RUN   TestWorkspacesUpdate/when_updating_a_subset_of_values
--- PASS: TestWorkspacesUpdate (1.39s)
    --- PASS: TestWorkspacesUpdate/when_updating_a_subset_of_values (0.18s)
=== RUN   TestWorkspacesUpdateByID
=== RUN   TestWorkspacesUpdateByID/when_updating_a_subset_of_values
--- PASS: TestWorkspacesUpdateByID (1.02s)
    --- PASS: TestWorkspacesUpdateByID/when_updating_a_subset_of_values (0.15s)
PASS
ok  	github.com/hashicorp/go-tfe	2.671s
?   	github.com/hashicorp/go-tfe/examples/organizations	[no test files]
?   	github.com/hashicorp/go-tfe/examples/workspaces	[no test files]
?   	github.com/hashicorp/go-tfe/mocks	[no test files]
```

File: organization.go [hunk 9]
```diff
@@ -68,6 +68,7 @@ type Organization struct {
 	CostEstimationEnabled                             bool                     `jsonapi:"attr,cost-estimation-enabled"`
 	CreatedAt                                         time.Time                `jsonapi:"attr,created-at,iso8601"`
 	Email                                             string                   `jsonapi:"attr,email"`
+	AssessmentsEnforced                               bool                     `jsonapi:"attr,assessments-enforced"`
```
Comment (annawinkler) [inline organization.go:71]: 💅 Could you add this attribute in alphabetical order?

File: workspace.go [hunk 10]
```diff
@@ -115,6 +115,7 @@ type Workspace struct {
 	CanQueueDestroyPlan        bool                  `jsonapi:"attr,can-queue-destroy-plan"`
 	CreatedAt                  time.Time             `jsonapi:"attr,created-at,iso8601"`
 	Description                string                `jsonapi:"attr,description"`
+	AssessmentsEnabled         bool                  `jsonapi:"attr,assessments-enabled"`
```
Comment (annawinkler) [inline workspace.go:118]: 💅 Could you add this attribute in alphabetical order?

**Review by annawinkler — APPROVED** [review]
🔥 🍪

**Review by annawinkler — APPROVED** [review]
:shipit:

**Pull request merged**